src/Init/Data/Format/Basic.lean

```diff
@@ -89,36 +88,35 @@ private partial def spaceUptoLine' : List WorkGroup → Nat → SpaceResult
   |   { items := [],    .. }::gs, w => spaceUptoLine' gs w
   | g@{ items := i::is, .. }::gs, w => merge w (spaceUptoLine i.f g.flatten w) (spaceUptoLine' ({ g with items := is }::gs))
 
-private structure State where
-  out    : String := ""
-  column : Nat    := 0
+/-- A monad in which we can pretty-print `Format` objects. -/
+class MonadPrettyFormat (m : Type → Type) where
+  pushOutput (s : String) : m Unit
+  pushNewline (indent : Nat) : m Unit
+  currColumn : m Nat
+  withTag {α : Type} (t : Nat) : m α → m α
+open MonadPrettyFormat
 
-private def pushGroup (flb : FlattenBehavior) (items : List WorkItem) (gs : List WorkGroup) (w : Nat) : StateM State (List WorkGroup) := do
-  let k  := (← get).column
+private def pushGroup (flb : FlattenBehavior) (items : List WorkItem) (gs : List WorkGroup) (w : Nat) [Monad m] [MonadPrettyFormat m] : m (List WorkGroup) := do
+  let k  ← currColumn
   -- Flatten group if it + the remainder (gs) fits in the remaining space. For `fill`, measure only up to the next (ungrouped) line break.
   let g  := { flatten := flb == FlattenBehavior.allOrNone, flb := flb, items := items : WorkGroup }
   let r  := spaceUptoLine' [g] (w-k)
   let r' := merge (w-k) r (spaceUptoLine' gs);
   -- Prevent flattening if any item contains a hard line break, except within `fill` if it is ungrouped (=> unflattened)
   return { g with flatten := !r.foundFlattenedHardLine && r'.space <= w-k }::gs
 
-private def pushOutput (s : String) : StateM State Unit :=
-  -- We avoid a structure instance update, and write this function using pattern matching because of issue #361
-  modify fun ⟨out, col⟩ => ⟨out ++ s, col + s.length⟩
-
-private def pushNewline (indent : Nat) : StateM State Unit :=
-  modify fun st => { st with out := st.out ++ "\n".pushn ' ' indent, column := indent }
-
-private partial def be (w : Nat) : List WorkGroup → StateM State Unit
+private partial def be (w : Nat) [Monad m] [MonadPrettyFormat m] : List WorkGroup → m Unit
   | []                           => pure ()
   |   { items := [],    .. }::gs => be w gs
   | g@{ items := i::is, .. }::gs => do
     let gs' (is' : List WorkItem) := { g with items := is' }::gs;
     match i.f with
     | nil => be w (gs' is)
-    | tag _ f => be w (gs' ({ i with f }::is))
+    | tag t f =>
+      withTag t <| be w [{ g with items := [{ i with f }] }]
```

Comment:
I believe this prevents `be` from being tail-recursive. However I observed no performance degradation due to this with a fairly long `#synth` trace taken from binport `Lib4`. If this is a worry we could consider specializing `be`. cc @leodemoura

Reply:
@Vtec234 I am not worried about performance degradation, but potential stack overflows.

Reply:
Fixed. I had to add an additional fake tag stack size to `WorkItem` but it is tail-recursive again. The default `MonadPrettyFormat` does nothing on tags so should hopefully specialize away.